torch/lib/c10d/TCPStore.cpp

```diff
@@ -57,11 +56,174 @@ TCPStoreDaemon::~TCPStoreDaemon() {
   closeStopSignal();
 }
 
-void TCPStoreDaemon::join() {
+void BackgroundThread::join() {
   daemonThread_.join();
 }
 
-void TCPStoreDaemon::queryFds(std::vector<struct pollfd>& fds) {
+#ifdef _WIN32
+void BackgroundThread::initStopSignal() {
+  ghStopEvent_ = CreateEvent(NULL, TRUE, FALSE, NULL);
+  if (ghStopEvent_ == NULL) {
+    throw std::runtime_error(
+        "Failed to create the control pipe to start the "
+        "BackgroundThread run");
+  }
+}
+
+void BackgroundThread::closeStopSignal() {
+  CloseHandle(ghStopEvent_);
+}
+
+void BackgroundThread::stop() {
+  SetEvent(ghStopEvent_);
+}
+#else
+void BackgroundThread::initStopSignal() {
+  if (pipe(controlPipeFd_.data()) == -1) {
+    throw std::runtime_error(
+        "Failed to create the control pipe to start the "
+        "BackgroundThread run");
+  }
+}
+
+void BackgroundThread::closeStopSignal() {
+  for (auto fd : controlPipeFd_) {
+    if (fd != -1) {
+      ::close(fd);
+    }
+  }
+}
+
+void BackgroundThread::stop() {
+  if (controlPipeFd_[1] != -1) {
+    // close the write end of the pipe
+    ::close(controlPipeFd_[1]);
+    controlPipeFd_[1] = -1;
+  }
+}
+#endif
+
+// TCPStoreListener class methods
+TCPStoreWorkerDaemon::TCPStoreWorkerDaemon(int listenSocket)
+    : BackgroundThread(listenSocket) {
+  daemonThread_ = std::thread(&TCPStoreWorkerDaemon::run, this);
+}
+
+void TCPStoreWorkerDaemon::addCallback(
+    std::string key,
+    std::function<void(c10::optional<std::string>, c10::optional<std::string>)>
+        callback) {
+  const std::lock_guard<std::mutex> lock(keyToCallbacksMutex);
+  keyToCallbacks[key] = callback;
+}
+
+// Runs all the callbacks that the worker has registered
+void TCPStoreWorkerDaemon::callbackHandler(int socket) {
+  auto watchResponse = tcputil::recvValue<WatchResponseType>(socket);
+  if (watchResponse == WatchResponseType::KEY_CALLBACK_REGISTERED) {
+    // Notify the waiting "watchKey" operation to return
+    callbackRegistered = true;
```

Comment:
this can be accessed by multiple threads? 

1. callbackHandler can set this to true
2. TCPStore::watchKey can set this to false concurrently?

Reply:
Great point, I will add a lock to TCPStore::watchkey